HNSW.h

```diff
@@ -0,0 +1,286 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD+Patents license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+// -*- c++ -*-
+
+#pragma once
+
+#include <vector>
+#include <unordered_set>
+#include <queue>
+
+#include <omp.h>
+
+#include "Index.h"
+#include "FaissAssert.h"
+#include "utils.h"
+
+
+namespace faiss {
+
+
+/** Implementation of the Hierarchical Navigable Small World
+ * datastructure.
+ *
+ * Efficient and robust approximate nearest neighbor search using
+ * Hierarchical Navigable Small World graphs
+ *
+ *  Yu. A. Malkov, D. A. Yashunin, arXiv 2017
+ *
+ * This implmentation is heavily influenced by the NMSlib
+ * implementation by Yury Malkov and Leonid Boystov
+ * (https://github.com/searchivarius/nmslib)
+ *
+ * The HNSW object stores only the neighbor link structure, see
+ * IndexHNSW below for the full index object.
+ */
+
+
+struct VisitedTable;
+
+
+struct HNSW {
+  /// internal storage of vectors (32 bits: this is expensive)
+  typedef int storage_idx_t;
+
+  /// Faiss results are 64-bit
```

Comment:
Since it's documented that all `idx` types are 64-bit, it might be worth changing the definition of `idx_t` to `int64_t` at some point. This should at least be more predictable to those trying to add support for other platforms.

Reply:
Hmmm. The first change I would like to have is moving all `long` to `idx_t`. `long` is actually 32-bit on Windows.